packages/core/src/app/shipping/ShippingAddressForm.tsx

```diff
@@ -117,25 +62,69 @@ class ShippingAddressForm extends Component<
             ? `customFields.${fieldName}`
             : fieldName;
 
-        setFieldValue(`${addressFieldName}.${formFieldName}`, fieldValue);
+        formikSetFieldValue(`${addressFieldName}.${formFieldName}`, fieldValue);
     };
 
-    private handleChange: (fieldName: string, value: string) => void = (fieldName, value) => {
-        const { onFieldChange } = this.props;
-
+    const handleChange = (fieldName: string, value: string) => {
         onFieldChange(fieldName, value);
     };
 
-    private handleAutocompleteToggle: (state: { inputValue: string; isOpen: boolean }) => void = ({
+    const handleAutocompleteToggle = ({
         isOpen,
         inputValue,
+    }: {
+        inputValue: string;
+        isOpen: boolean;
     }) => {
-        const { onFieldChange } = this.props;
-
         if (!isOpen) {
             onFieldChange('address1', inputValue);
         }
     };
-}
+
+    const hasAddresses = addresses && addresses.length > 0;
+    const hasValidCustomerAddress = isValidCustomerAddress(
+        shippingAddress,
+        addresses,
+        formFields,
+    );
+
+    return (
+        <Fieldset id="checkoutShippingAddress">
+            {hasAddresses && (
+                <Fieldset id="shippingAddresses">
+                    <LoadingOverlay isLoading={isLoading}>
+                        <AddressSelect
+                            addresses={addresses}
+                            onSelectAddress={onAddressSelect}
+                            onUseNewAddress={onUseNewAddress}
+                            selectedAddress={
+                                hasValidCustomerAddress ? shippingAddress : undefined
+                            }
+                            type={AddressType.Shipping}
+                        />
+                    </LoadingOverlay>
+                </Fieldset>
+            )}
+
+            {!hasValidCustomerAddress && (
+                <LoadingOverlay isLoading={isLoading} unmountContentWhenLoading>
+                    <AddressForm
+                        countries={countries}
+                        countriesWithAutocomplete={countriesWithAutocomplete}
+                        countryCode={formAddress && formAddress.countryCode}
+                        fieldName={addressFieldName}
+                        formFields={formFields}
+                        googleMapsApiKey={googleMapsApiKey}
+                        isFloatingLabelEnabled={isFloatingLabelEnabled}
+                        onAutocompleteToggle={handleAutocompleteToggle}
+                        onChange={handleChange}
+                        setFieldValue={setFieldValue}
+                        shouldShowSaveAddress={shouldShowSaveAddress}
+                    />
+                </LoadingOverlay>
+            )}
+        </Fieldset>
+    );
+};
 
 export default connectFormik(ShippingAddressForm);
```

Comment:
As discussed offline, we decided to keep Formik HOC.

Reply:
I have removed 3 props.

Not reducing prop drilling completely because:
1. Some props are expensive. For example, `address`, we may consider calculate it once and put it in a context:

   ```ts
   const shippableItemsCount = getShippableItemsCount(cart);
   const methodId = getShippingMethodId(checkout, config);
   const shouldShowMultiShipping =
       hasMultiShippingEnabled && !methodId && shippableItemsCount > 1;

   const shippingAddress =
       !shouldShowMultiShipping && consignments.length > 1 ? undefined : getShippingAddress();
    ```
2.	Lower-level component isn’t a function component yet, so props are still needed.
For example, `googleMapsApiKey` should be derived in `<AddressForm />`, not here.